File: week1-python/Image_Annotation/Drawing_Text.py
# Import cv2 module
import cv2
from dataPath import DATA_PATH

# Path to the image we are going to read
# This can be an absolute or relative path
# Here we are using a relative path
imageName = DATA_PATH+"images/boy.jpg"

# Load the image
image = cv2.imread(imageName, cv2.IMREAD_COLOR)

# Put text into image
imageText = image.copy()
text = "I am studying"
fontScale = 1.5
fontFace = cv2.FONT_HERSHEY_COMPLEX
fontColor = (250, 10, 10)
fontThickness = 2
cv2.putText(imageText, text, (20, 350), fontFace, fontScale, fontColor, fontThickness, cv2.LINE_AA);

# Display the image
#cv2.imshow("Image Text",imageText)
#cv2.waitKey(0)
cv2.imwrite("../results/imageText.png",imageText)

#Issues in Text annotation

#Solution 1: Get font size from pixel height of text

pixelHeight = 20

# Calculate the fontScale
fontScale = cv2.getFontScaleFromHeight(fontFace, pixelHeight, fontThickness)
print("fontScale = {}".format(fontScale))

imageTextFontScale = image.copy()
cv2.putText(imageTextFontScale, text, (20, 350), fontFace, fontScale, fontColor, fontThickness, cv2.LINE_AA);
# Display the image
#cv2.imshow("Image Text FontScale",imageTextFontScale)
#cv2.waitKey(0)
cv2.imwrite("../results/imageTextFontScale.png",imageTextFontScale)

#Solution 2: Get height and width of text

imageGetTextSize = image.copy()
imageHeight, imageWidth=imageGetTextSize.shape[:2]

# Get the text box height and width and also the baseLine
textSize, baseLine = cv2.getTextSize(text,fontFace,fontScale,fontThickness)
textWidth,textHeight = textSize
print("TextWidth = {}, TextHeight = {}, baseLine = {}".format(textWidth, textHeight, baseLine))

# Get the coordinates of text box bottom left corner
# The xccordinate will be such that the text is centered
xcoordinate = (imageWidth - textWidth)//2
# The y coordinate will be such that the entire box is just 10 pixels above the bottom of image
ycoordinate = (imageHeight - baseLine - 10)
print("TextBox Bottom Left = ({},{})".format(xcoordinate,ycoordinate))

# Draw the Canvas using a filled rectangle
canvasColor = (255, 255, 255)
canvasBottomLeft = (xcoordinate,ycoordinate+baseLine)
canvasTopRight = (xcoordinate+textWidth, ycoordinate-textHeight)
cv2.rectangle(imageGetTextSize, canvasBottomLeft, canvasTopRight, canvasColor, thickness=-1);
print("Canvas Bottom Left = {}, Top Right = {}".format(canvasBottomLeft,canvasTopRight))

# Now draw the baseline ( just for reference )
lineThickness = 2
lineLeft = (xcoordinate, ycoordinate)
lineRight = (xcoordinate+textWidth, ycoordinate)
lineColor = (0,255,0)
cv2.line(imageGetTextSize, lineLeft, lineRight, lineColor, thickness = lineThickness, lineType=cv2.LINE_AA);

# Finally Draw the text
cv2.putText(imageGetTextSize, text, (xcoordinate,ycoordinate), fontFace, fontScale, fontColor, fontThickness, cv2.LINE_AA);

# Display the Output Image
#cv2.imshow("Image Get TextSize",imageGetTextSize)
#cv2.waitKey(0)
cv2.imwrite("../results/imageGetTextSize.png",imageGetTextSize)

File: week1-cpp/Image_Annotation/Drawing_Text.cpp
// Include libraries
#include <iostream>
#include "dataPath.hpp"
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
using namespace std;
using namespace cv;

int main(void){

    // Path to the image we are going to read
    // This can be an absolute or relative path
    // Here we are using a relative path
    string imageName = DATA_PATH+"images/boy.jpg";

    // Load the image
    Mat image;
    image = imread(imageName, IMREAD_COLOR);

	// Put text on image
	Mat imageText = image.clone();
	string text = "I am studying";
	double fontScale = 1.5;
	int fontFace = FONT_HERSHEY_SIMPLEX;
	Scalar fontColor = Scalar(250, 10, 10);
	int fontThickness = 2;

	putText(imageText, text, Point(20, 350), fontFace, fontScale, fontColor, fontThickness, LINE_AA);

	imwrite("../results/text.jpg",imageText);
	//imshow("Drawing text on image", imageText);
	//waitKey(0);
	
	int pixelHeight = 20;

	// Calculate the fontScale
	fontScale = getFontScaleFromHeight(fontFace, pixelHeight, fontThickness);
	cout << "fontScale = " << fontScale << endl;
	
	Mat imageTextFontScale;
	imageTextFontScale = image.clone();
	putText(imageTextFontScale, text, Point(20, 350), fontFace, fontScale, fontColor, fontThickness, LINE_AA);

	imwrite("../results/text2.jpg",imageTextFontScale);
	//imshow("Text written using fontScale",imageTextFontScale);
	//waitKey(0);
	
	Mat imageGetTextSize;
	imageGetTextSize = image.clone();
	int imageHeight = imageGetTextSize.rows;
	int imageWidth = imageGetTextSize.cols;

	// Get the text box height and width and also the baseLine
	int baseLine = 0;
	Size textSize = getTextSize(text,fontFace,fontScale,fontThickness, &baseLine);
	int textWidth = textSize.width;
	int textHeight = textSize.height;

	cout << "TextWidth = " << textWidth << ", TextHeight = " << textHeight << ", baseLine = " << baseLine << endl;

	// Get the coordinates of text box bottom left corner
	// The xccordinate will be such that the text is centered
	int xcoordinate = (imageWidth - textWidth)/2;
	// The y coordinate will be such that the entire box is just 10 pixels above the bottom of image
	int ycoordinate = (imageHeight - baseLine - 10);

	cout << "TextBox Bottom Left = (" << xcoordinate << "," << ycoordinate << ")" << endl;

	// Draw the Canvas using a filled rectangle
	Scalar canvasColor = Scalar(255, 255, 255);
	Point canvasBottomLeft (xcoordinate,ycoordinate+baseLine);
	Point canvasTopRight (xcoordinate+textWidth, ycoordinate-textHeight);
	
	rectangle(imageGetTextSize, canvasBottomLeft, canvasTopRight, canvasColor, -1);

	cout << "Canvas Bottom Left = " << canvasBottomLeft << ", Top Right = " << canvasTopRight << endl;

	// Now draw the baseline ( just for reference )
	int lineThickness = 2;
	Point lineLeft (xcoordinate, ycoordinate);
	Point lineRight (xcoordinate+textWidth, ycoordinate);
	Scalar lineColor = Scalar(0,255,0);

	line(imageGetTextSize, lineLeft, lineRight, lineColor, lineThickness, LINE_AA);

	// Finally Draw the text
	putText(imageGetTextSize, text, Point(xcoordinate,ycoordinate), fontFace, fontScale, fontColor, fontThickness, LINE_AA);

	imwrite("../results/text3.jpg",imageGetTextSize);

	//imshow("Write image using calculated text size",imageGetTextSize);
	//waitKey(0);
	return 0;
}
